pkg/batcher/batch_builder.go

```diff
@@ -0,0 +1,173 @@
+package batcher
+
+import (
+	"github.com/StackVista/stackstate-agent/pkg/collector/check"
+	"github.com/StackVista/stackstate-agent/pkg/health"
+	"github.com/StackVista/stackstate-agent/pkg/topology"
+)
+
+// CheckInstanceBatchState is the type representing batched data per check instance
+type CheckInstanceBatchState struct {
+	Topology *topology.Topology
+	Health   map[string]health.Health
+}
+
+// CheckInstanceBatchStates is the type representing batched data for all check instances
+type CheckInstanceBatchStates map[check.ID]CheckInstanceBatchState
+
+// BatchBuilder is a helper class to build Topology based on submitted data, this data structure is not thread safe
+type BatchBuilder struct {
+	states CheckInstanceBatchStates
+	// Count the amount of elements we gathered
+	elementCount int
+	// Amount of elements when we flush
+	maxCapacity int
+}
+
+// NewBatchBuilder constructs a BatchBuilder
+func NewBatchBuilder(maxCapacity int) BatchBuilder {
+	return BatchBuilder{
+		states:       make(map[check.ID]CheckInstanceBatchState),
+		elementCount: 0,
+		maxCapacity:  maxCapacity,
+	}
+}
+
+func (builder *BatchBuilder) getState(checkID check.ID) CheckInstanceBatchState {
+	if value, ok := builder.states[checkID]; ok {
+		return value
+	}
+
+	state := CheckInstanceBatchState{
+		Topology: nil,
+		Health:   make(map[string]health.Health),
+	}
+	builder.states[checkID] = state
+	return state
+}
+
+func (builder *BatchBuilder) getTopology(checkID check.ID, instance topology.Instance) *topology.Topology {
+	state := builder.getState(checkID)
+
+	if state.Topology != nil {
+		return state.Topology
+	}
+
+	builder.states[checkID] = CheckInstanceBatchState{
+		Topology: &topology.Topology{
+			StartSnapshot: false,
+			StopSnapshot:  false,
+			Instance:      instance,
+			Components:    make([]topology.Component, 0),
+			Relations:     make([]topology.Relation, 0),
+		},
+		Health: state.Health,
+	}
+	return builder.states[checkID].Topology
+}
+
+func (builder *BatchBuilder) getHealth(checkID check.ID, stream health.Stream) health.Health {
+	state := builder.getState(checkID)
+
+	if value, ok := state.Health[stream.GoString()]; ok {
+		return value
+	}
+
+	builder.states[checkID] = CheckInstanceBatchState{
+		Topology: state.Topology,
+		Health:   make(map[string]health.Health),
```

Comment:
here you are re-initializing the health, but what if you have other healths already ?

Reply:
done